src/Sylius/Bundle/ShopBundle/Grid/ProductGrid.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ShopBundle\Grid;
+
+use Sylius\Bundle\GridBundle\Builder\Field\DateTimeField;
+use Sylius\Bundle\GridBundle\Builder\Field\Field;
+use Sylius\Bundle\GridBundle\Builder\Field\StringField;
+use Sylius\Bundle\GridBundle\Builder\Filter\StringFilter;
+use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;
+use Sylius\Bundle\GridBundle\Grid\AbstractGrid;
+use Sylius\Bundle\GridBundle\Grid\ResourceAwareGridInterface;
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Core\Model\TaxonInterface;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Webmozart\Assert\Assert;
+
+final class ProductGrid extends AbstractGrid implements ResourceAwareGridInterface
+{
+    /**
+     * @param TaxonRepositoryInterface<TaxonInterface> $taxonRepository
+     */
+    public function __construct(
+        private string $resourceClass,
+        private ChannelContextInterface $channelContext,
+        private TaxonRepositoryInterface $taxonRepository,
+        private LocaleContextInterface $localeContext,
+        private RequestStack $requestStack,
+        private bool  $includeAllDescendants,
+    ) {
+    }
+
+    public static function getName(): string
+    {
+        return 'sylius_shop_product';
+    }
+
+    public function buildGrid(GridBuilderInterface $gridBuilder): void
+    {
+        $request = $this->requestStack->getMainRequest();
+        Assert::notNull($request, 'The '.self::class. ' does not work in CLI mode');
+
+        $localeCode = $this->localeContext->getLocaleCode();
+
+        // @see Sylius\Bundle\ResourceBundle\ExpressionLanguage\NotNullExpressionFunctionProvider
+        $taxon = $this->taxonRepository->findOneBySlug($request->attributes->get('slug'), $localeCode);
```

Comment:
Bad phrasing, it means "out of my head without checking or thinking about it too hard"

Reply:
Problem with `$request->attributes->get('slug')` it will return `null`, if current request doesn't have this slug, which means repository needs to handle `?string $slug`, which only possible right now because repository wasn't updated to 2.0 (Not type-hinted yet and this code will crash later in 2.0 upgrade).

If class has requirement, then it should assert for `slug` presence in request explicitly.

```php
$slug = $request->attributes->getString('slug');
Assert::notEmpty($slug, 'Request attributes slug must be set/preset/not empty');

$taxon = $this->taxonRepository->findOneBySlug($slug, $localeCode);
```

Reply:
This kind of situation will be repeating with yaml->php config conversion as it was not possible to write complete code in yaml with null and strict type checks.

Reply:
Thanks for catching that but I will look out for those kinds of situations in future migrations.